packages/react-debug-tools/src/ReactDebugHooks.js

```diff
@@ -495,16 +495,50 @@ function useFormState<S, P>(
   const hook = nextHook(); // FormState
   nextHook(); // ActionQueue
   let state;
+  let debugInfo = null;
   if (hook !== null) {
-    state = hook.memoizedState;
+    const actionResult = hook.memoizedState;
+    if (
+      typeof actionResult === 'object' &&
+      actionResult !== null &&
+      // $FlowFixMe[method-unbinding]
+      typeof actionResult.then === 'function'
+    ) {
+      const thenable: Thenable<Awaited<S>> = (actionResult: any);
+      switch (thenable.status) {
+        case 'fulfilled': {
+          state = thenable.value;
+          debugInfo =
+            thenable._debugInfo === undefined ? null : thenable._debugInfo;
+          break;
+        }
+        case 'rejected': {
+          const rejectedError = thenable.reason;
+          throw rejectedError;
+        }
+        default:
+          // If this was an uncached Promise we have to abandon this attempt
+          // but we can still emit anything up until this point.
+          hookLog.push({
+            primitive: 'Unresolved',
```

Comment:
This needs a different name and you need to exhaust this case in the stack frame cache. Otherwise this will generate the wrong stack frame information and so it'll generate a messed up tree.

Probably should add some tests for it.

Reply:
Did that just like use but it's not clear to me how I could test it. It was working fine in the fixture without the case as far as I could tell. 

I guess I would need to simulate an update in `getPrimitiveStackCache` because we never actually call `useFormState` with a thenable?

Reply:
> This needs a different name

Can it just use `FormState`? I don't understand the purpose of using different primitive names for the same hook.

Reply:
It's annoying because this is used both for the key in the cache map and each one will have different stack traces. So it needs to be unique per stack. However, it's also used for the display name.

Which is why I gave it a unique primitive but then aliased it back:

https://github.com/facebook/react/blob/main/packages/react-debug-tools/src/ReactDebugHooks.js#L795

We could just make two separate fields for the name and the key potentially.

Since both your cases are in the same Hook you can also just unify them so that they have the same `new Error().stack` and ideally the same `hookLog.push(...)`.

Reply:
Unified stack error and hook log. I have to fight flow analysis more when unifying the hook log push more unless I tie up `value` and `error` into a single type like `{ value: Awaited<S>, error: null } | { value: Thenable, error: mixed }`.